### PR TITLE
Rule system for hiding/displaying fields based on previous input

### PR DESCRIPTION
**Pull request by akaivola**

### CODE REVIEW

**Comment by auramo** [comment]
It's not obvious in which form to write the time of birth.
For example this is ok:
01.1.2011
But this is not:
1.1.2011

Should be show a HTML placeholder giving an example?
(http://www.w3schools.com/tags/att_input_placeholder.asp)

**Comment by auramo** [comment]
Server-side validation allows this (nationality: Suomi but no ssn, only birth-date)

` curl -v 'http://localhost:8351/hakemus/api/application' -H 'Content-Type: application/json; charset=UTF-8' -H 'Cookie: _hjUserId=e4ad65c7-abe5-3e61-8177-413fb3d76101; _ga=GA1.1.501889459.1454313078; ring-session=bab43770-b935-445f-b2ec-313685392e6c' --data-binary '{"form":5,"lang":"fi","answers":[{"key":"address","value":"katuyis","fieldType":"textField","label":{"fi":"Katuosoite","sv":"Adress"}},{"key":"email","value":"a@a.com","fieldType":"textField","label":{"fi":"S\u00e4hk\u00f6postiosoite","sv":"E-postadress"}},{"key":"preferred-name","value":"a","fieldType":"textField","label":{"fi":"Kutsumanimi","sv":"Smeknamn"}},{"key":"last-name","value":"a","fieldType":"textField","label":{"fi":"Sukunimi","sv":"Efternamn"}},{"key":"phone","value":"040484888","fieldType":"textField","label":{"fi":"Matkapuhelin","sv":"Mobiltelefonnummer"}},{"key":"nationality","value":"Suomi","fieldType":"dropdown","label":{"fi":"Kansalaisuus","sv":"Nationalitet"}},{"key":"first-name","value":"a","fieldType":"textField","label":{"fi":"Etunimet","sv":"F\u00f6rnamn"}},{"key":"birth-date","value":"01.1.2013","fieldType":"textField","label":{"fi":"Syntym\u00e4aika","sv":"F\u00f6delsedag"}},{"key":"postal-code","value":"00730","fieldType":"textField","label":{"fi":"Postinumero","sv":"Postnummer"}},{"key":"language","value":"suomi","fieldType":"dropdown","label":{"fi":"\u00c4idinkieli","sv":"Modersm\u00e5l"}},{"key":"gender","value":"Mies","fieldType":"dropdown","label":{"fi":"Sukupuoli","sv":"K\u00f6n"}},{"key":"postal-office","value":"postitmp","fieldType":"textField","label":{"fi":"Postitoimipaikka","sv":"Postkontor"}},{"key":"home-town","value":"helsinbki","fieldType":"textField","label":{"fi":"Kotikunta","sv":"Bostadsort"}}]}' --compressed`

**Comment by auramo** [comment]
Tests do not pass:

`CONFIG=config/test.edn lein spec -t unit
`
->
`  1) java.lang.Exception: Oops!  It looks like you tried to add 'nil' to a spec.  That's probabaly not what you wanted., compiling:(/Users/auramo/src/ataru/spec/ataru/hakija/application_validators_spec.clj:76:1)
`

File: src/cljc/ataru/schema/form_schema.cljc [hunk 1]
```diff
@@ -50,6 +50,7 @@
 (s/defschema FormField {:fieldClass (s/eq "formField")
                         :id s/Str
                         (s/optional-key :validators) [(apply s/enum (keys validator/validators))]
+                        (s/optional-key :rules) {s/Keyword s/Any}
```
Comment (auramo) [inline src/cljc/ataru/schema/form_schema.cljc]: keyword and any?

File: src/cljc/ataru/hakija/application_validators.cljc [hunk 2]
```diff
@@ -53,36 +57,71 @@
...
 
-(def validators {"required"    required
-                 "ssn"         ssn
-                 "email"       email
-                 "postal-code" postal-code
-                 "phone"       phone})
+#?(:clj
+   (def parse-date
+     (let [formatter (f/formatter (c/time-zone-for-id "Europe/Helsinki")
+                                  "dd.MM.YYYY"
```
Comment (akaivola) [inline src/cljc/ataru/hakija/application_validators.cljc]: parse date with d.M.YYYY

**Pull request merged**